pkg/cmd/roachtest/roachtestutil/clusterupgrade/clusterupgrade.go

```diff
@@ -156,10 +157,22 @@ func LatestPatchRelease(series string) (*Version, error) {
 // associated with the given database connection.
 // NB: version means major.minor[-internal]; the patch level isn't
 // returned. For example, a binary of version 19.2.4 will return 19.2.
-func BinaryVersion(ctx context.Context, db *gosql.DB) (roachpb.Version, error) {
+func BinaryVersion(ctx context.Context, l *logger.Logger, db *gosql.DB) (roachpb.Version, error) {
 	zero := roachpb.Version{}
 	var sv string
-	if err := db.QueryRowContext(ctx, `SELECT crdb_internal.node_executable_version();`).Scan(&sv); err != nil {
+	rows, err := roachtestutil.QueryWithRetry(
```

Comment:
You could have kept the return type as `*Row`; `Scan` already handles error cases, so you don't need do it explicitly. For the other call site in `CheckInvalidDescriptors`, you just unpack the rows and err from the `Row` struct.